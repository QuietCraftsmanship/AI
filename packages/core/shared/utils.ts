import { customAlphabet } from 'nanoid/non-secure'
import { JSONValue } from './types'

// 7-character random string
export const nanoid = customAlphabet(
  '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz',
  7
)

export function createChunkDecoder() {
  const decoder = new TextDecoder()
  return function (chunk: Uint8Array | undefined): string {
    if (!chunk) return ''
    return decoder.decode(chunk, { stream: true })
  }
}

/**
 * The map of prefixes for data in the stream
 *
 * - 0: Text from the LLM response
 * - 1: (OpenAI) function_call responses
 * - 2: custom JSON added by the user using `Data`
 *
 * Example:
 * ```
 * 0:Vercel
 * 0:'s
 * 0: AI
 * 0: AI
 * 0: SDK
 * 0: is great
 * 0:!
 * 2: { "someJson": "value" }
 * 1: {"function_call": {"name": "get_current_weather", "arguments": "{\\n\\"location\\": \\"Charlottesville, Virginia\\",\\n\\"format\\": \\"celsius\\"\\n}"}}
 *```
 */
export const StreamStringPrefixes = {
  text: 0,
  function_call: 1,
  data: 2
} as const

/**
 * Prepends a string with a prefix from the `StreamChunkPrefixes`, JSON-ifies it, and appends a new line.
 */
export const getStreamString = (
  type: keyof typeof StreamStringPrefixes,
  value: JSONValue
): StreamString =>
  `${StreamStringPrefixes[type]}:${
    typeof value === 'string' ? value : JSON.stringify(value)
  }\n`

export type StreamString =
  `${(typeof StreamStringPrefixes)[keyof typeof StreamStringPrefixes]}:${string}\n`
