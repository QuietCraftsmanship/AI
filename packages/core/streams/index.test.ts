import { setup } from '../tests/utils/mock-service'
import { OpenAIStream, StreamingTextResponse } from '.'
import { createClient } from '../tests/utils/mock-client'

describe('AIStream', () => {
  let server: ReturnType<typeof setup>
  beforeAll(() => {
    server = setup()
  })
  afterAll(() => {
    server.teardown()
  })

  describe('OpenAIStream', () => {
    it('should be able to parse SSE and receive the streamed response', async () => {
      const stream = OpenAIStream(
        await fetch(server.api, {
          headers: {
            'x-mock-service': 'openai',
            'x-mock-type': 'chat'
          }
        })
      )
      const response = new StreamingTextResponse(stream)
      const client = createClient(response)
      const chunks = await client.readAll()
      expect(chunks).toMatchSnapshot()
      expect(server.getRecentFlushed()).toMatchSnapshot()
    })

    it('should handle backpressure on the server', async () => {
      const controller = new AbortController()
      const stream = OpenAIStream(
        await fetch(server.api, {
          headers: {
            'x-mock-service': 'openai',
            'x-mock-type': 'chat'
          },
          signal: controller.signal
        })
      )
      const response = new StreamingTextResponse(stream)
      const client = createClient(response)
      const chunks = await client.readAndAbort(controller)
      expect(chunks).toMatchSnapshot()
      expect(server.getRecentFlushed()).toMatchSnapshot()
    })
  })
})
