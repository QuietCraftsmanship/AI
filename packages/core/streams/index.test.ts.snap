// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AIStream OpenAIStream should be able to parse SSE and receive the streamed response 1`] = `
[
  "Hello",
  ",",
  " world",
  ".",
]
`;

exports[`AIStream OpenAIStream should be able to parse SSE and receive the streamed response 2`] = `
[
  {
    "choices": [
      {
        "delta": {
          "role": "assistant",
        },
        "finish_reason": null,
        "index": 0,
      },
    ],
    "created": 1686901302,
    "id": "chatcmpl-7RyNSW2BXkOQQh7NlBc65j5kX8AjC",
    "model": "gpt-3.5-turbo-0301",
    "object": "chat.completion.chunk",
  },
  {
    "choices": [
      {
        "delta": {
          "content": "Hello",
        },
        "finish_reason": null,
        "index": 0,
      },
    ],
    "created": 1686901302,
    "id": "chatcmpl-7RyNSW2BXkOQQh7NlBc65j5kX8AjC",
    "model": "gpt-3.5-turbo-0301",
    "object": "chat.completion.chunk",
  },
  {
    "choices": [
      {
        "delta": {
          "content": ",",
        },
        "finish_reason": null,
        "index": 0,
      },
    ],
    "created": 1686901302,
    "id": "chatcmpl-7RyNSW2BXkOQQh7NlBc65j5kX8AjC",
    "model": "gpt-3.5-turbo-0301",
    "object": "chat.completion.chunk",
  },
  {
    "choices": [
      {
        "delta": {
          "content": " world",
        },
        "finish_reason": null,
        "index": 0,
      },
    ],
    "created": 1686901302,
    "id": "chatcmpl-7RyNSW2BXkOQQh7NlBc65j5kX8AjC",
    "model": "gpt-3.5-turbo-0301",
    "object": "chat.completion.chunk",
  },
  {
    "choices": [
      {
        "delta": {
          "content": ".",
        },
        "finish_reason": null,
        "index": 0,
      },
    ],
    "created": 1686901302,
    "id": "chatcmpl-7RyNSW2BXkOQQh7NlBc65j5kX8AjC",
    "model": "gpt-3.5-turbo-0301",
    "object": "chat.completion.chunk",
  },
  {
    "choices": [
      {
        "delta": {},
        "finish_reason": "stop",
        "index": 0,
      },
    ],
    "created": 1686901302,
    "id": "chatcmpl-7RyNSW2BXkOQQh7NlBc65j5kX8AjC",
    "model": "gpt-3.5-turbo-0301",
    "object": "chat.completion.chunk",
  },
]
`;

exports[`AIStream OpenAIStream should handle backpressure on the server 1`] = `
[
  "Hello",
]
`;

exports[`AIStream OpenAIStream should handle backpressure on the server 2`] = `
[
  {
    "choices": [
      {
        "delta": {
          "role": "assistant",
        },
        "finish_reason": null,
        "index": 0,
      },
    ],
    "created": 1686901302,
    "id": "chatcmpl-7RyNSW2BXkOQQh7NlBc65j5kX8AjC",
    "model": "gpt-3.5-turbo-0301",
    "object": "chat.completion.chunk",
  },
  {
    "choices": [
      {
        "delta": {
          "content": "Hello",
        },
        "finish_reason": null,
        "index": 0,
      },
    ],
    "created": 1686901302,
    "id": "chatcmpl-7RyNSW2BXkOQQh7NlBc65j5kX8AjC",
    "model": "gpt-3.5-turbo-0301",
    "object": "chat.completion.chunk",
  },
]
`;
